core/templates/dev/head/components/forms/schema_editors/SchemaBasedListEditorDirective.js

```diff
@@ -135,19 +135,35 @@ oppia.directive('schemaBasedListEditor', [
             }
           };
 
+          var deleteEmptyElements = function() {
+            for (var i = 0; i < $scope.localValue.length - 1; i++) {
+              if ($scope.localValue[i].length === 0) {
+                $scope.deleteElement(i);
+                i--;
+              }
+            }
+          };
+
           $scope.lastElementOnBlur = function() {
             _deleteLastElementIfUndefined();
             $scope.showAddItemButton();
           };
 
           $scope.showAddItemButton = function() {
+            deleteEmptyElements();
             $scope.isAddItemButtonPresent = true;
           };
 
           $scope.hideAddItemButton = function() {
             $scope.isAddItemButtonPresent = false;
           };
 
+          if ($scope.itemSchema().value === 'setInput') {
```

Comment:
Btw, for future reference, one way to tell that this is a "code smell" is that it refers to a concept outside its domain of responsibility. This is a just form for displaying lists. It should not know about "setInput".